firedrake/slate/preconditioners.py

```diff
@@ -98,19 +98,21 @@ def initialize(self, pc):
 
         shapes = (V[self.vidx].finat_element.space_dimension(),
                   np.prod(V[self.vidx].shape))
-        weight_kernel = """
-        for (int i=0; i<%d; ++i) {
-        for (int j=0; j<%d; ++j) {
-        w[i][j] += 1.0;
-        }}""" % shapes
-
+        domain = "{[i,j]: 0 <= i < %d and 0 <= j < %d}" % shapes
+        instructions = """
+        for i, j
+            w[i,j] = w[i,j] + 1
+        end
+        """
         self.weight = Function(V[self.vidx])
-        par_loop(weight_kernel, ufl.dx, {"w": (self.weight, INC)})
-        self.average_kernel = """
-        for (int i=0; i<%d; ++i) {
-        for (int j=0; j<%d; ++j) {
-        vec_out[i][j] += vec_in[i][j]/w[i][j];
-        }}""" % shapes
+        par_loop(domain, instructions, ufl.dx, {"w": (self.weight, INC)})
+
+        instructions = """
+        for i, j
+            vec_out[i,j] = vec_out[i,j] + vec_in[i,j]/w[i,j]
+        end
```

Comment:
It's not super important for these kernels, but I guess it would be good to ensure the loop priority is somehow right. Does loopy's scheduler prioritise "stride-1" over something else? @inducer ?

Reply:
There is no built-in preference. Loop order is user-specified.